curvefs/src/mds/schedule/copysetSchedule.cpp

```diff
@@ -0,0 +1,267 @@
+        oneRoundGenOp += CopySetScheduleForPool(poolId);
+    }
+
+    LOG(INFO) << "copysetScheduler generate " << oneRoundGenOp
+              << " operators at this round";
+    return oneRoundGenOp;
+}
+
+void CopySetScheduler::GetCopySetDistribution(
+    const std::vector<CopySetInfo> &copysetList,
+    const std::vector<MetaServerInfo> &metaserverList,
+    std::map<MetaServerIdType, std::vector<CopySetInfo>> *out) {
+    // calculate the copysetlist by traversing the copyset list
+    for (auto item : copysetList) {
+        for (auto peer : item.peers) {
+            if (out->find(peer.id) == out->end()) {
+                (*out)[peer.id] = std::vector<CopySetInfo>{item};
+            } else {
+                (*out)[peer.id].emplace_back(item);
+            }
+        }
+    }
+
+    // remove offline chunkserver, and report empty list for empty chunkserver
+    for (auto item : metaserverList) {
+        // remove offline chunkserver
+        if (item.IsOffline()) {
+            out->erase(item.info.id);
+            continue;
+        }
+
+        // report empty list for chunkserver with no copyset
+        if (out->find(item.info.id) == out->end()) {
+            (*out)[item.info.id] = std::vector<CopySetInfo>{};
+        }
+    }
+}
+
+bool CopySetScheduler::TransferCopyset(const CopySetInfo &copyset,
+                                       MetaServerIdType sourceId,
+                                       MetaServerIdType destId) {
+    // generate operator
+    Operator op = operatorFactory.CreateChangePeerOperator(
+        copyset, sourceId, destId, OperatorPriority::NormalPriority);
+    op.timeLimit = std::chrono::seconds(changeTimeSec_);
+    LOG(INFO) << "copyset scheduler gen " << op.OpToString() << " on "
+              << copyset.CopySetInfoStr();
+
+    // add operator
+    if (!opController_->AddOperator(op)) {
+        LOG(INFO) << "copysetSchduler add op " << op.OpToString()
+                  << " fail, copyset has already has operator"
+                  << " or operator num exceeds the limit.";
+        return false;
+    }
+
+    // create copyset
+    if (!topo_->CreateCopySetAtMetaServer(copyset.id, destId)) {
+        LOG(ERROR) << "copysetScheduler create " << copyset.CopySetInfoStr()
+                   << " on metaServer: " << destId << " error, delete operator"
+                   << op.OpToString();
+        opController_->RemoveOperator(copyset.id);
+        return false;
+    }
+
+    LOG(INFO) << "copysetScheduler create " << copyset.CopySetInfoStr()
+              << "on metaServer:" << destId
+              << " success. generator op: " << op.OpToString() << "success";
+    return true;
+}
+
+bool CopySetScheduler::IsCopysetCanTransfer(const CopySetInfo &copyset) {
+    // return false if operator exists on copyset
+    Operator exist;
+    if (opController_->GetOperatorById(copyset.id, &exist)) {
+        return false;
+    }
+
+    // return false if copyset already has candidate
```

Comment:
Why copyset can not transfer if has candidate?

Reply:
If copyset has candidate, it means the copyset has some operation(add, remove, change peer) on it. It can't do new op until no candidate.